### DISCUSSION TITLE
Can we make train_X and train_Y optional for SingleTaskGP?

### DISCUSSION THREAD

**Discussion by dynobo** [meta-pytorch/botorch#1217] [Q&A]
The two init args `train_X`, `train_Y` of `SingleTaskGP` are [typed as `Tensor` (none-optional)](https://github.com/pytorch/botorch/blob/663d4e322485aac179163a28fcca6d0f2958f00e/botorch/models/gp_regression.py#L65-L66). It's parent class `gpytorch.models.ExactGP` actually [handles the case of `None` as input data](https://github.com/cornellius-gp/gpytorch/blob/master/gpytorch/models/exact_gp.py#L64-L69) and `SingleTaskGP` also seems to work well when instantiated with `train_X=None, train_Y=None`.

As it is sometimes useful to initialize _without_ training data and the implementation of the parent `ExactGP` is explicitly implemented to support that, does anything speak against typing them as `Optional[Tensor]` in botorch? 

(If not, I'd open a PR accordingly)

**Comment by saitcakmak** [comment]
Using `None` inputs will error out if you use input / outcome transforms. You can initialize it with empty tensors though.
```
model = SingleTaskGP(torch.ones(0, 2), torch.ones(0, 2), outcome_transform=Standardize(m=2))
```
(swapping for `None` above will error out)

**Reply by saitcakmak** [reply]
Looks like this also errors out with input transforms:
```
model = SingleTaskGP(torch.ones(0, 2), torch.ones(0, 2), outcome_transform=Standardize(m=2), input_transform=Normalize(d=2))

IndexError: min(): Expected reduction dim 0 to have non-zero size.

```

**Reply by Balandat** [reply]
@saitcakmak Could we just patch the `transform_inputs` on the `Model` to pass through `None`? Seems like that's the thing to do if what's of interest is the GP Prior.

**Reply by saitcakmak** [reply]
> Could we just patch the transform_inputs on the Model to pass through None?

That gets you through the model initialization with empty tensors. Sampling from the posterior still throws errors though:
```
empty_inputs = torch.ones(0, 2)
kwargs = {
    "train_X": empty_inputs,
    "train_Y": empty_inputs,
    "outcome_transform": Standardize(m=2),
    "input_transform": Normalize(d=2),
}
model = SingleTaskGP(**kwargs)
posterior = model.posterior(torch.rand(3, 4, 2))
samples = posterior.rsample(sample_shape=torch.Size([5]))
self.assertEqual(samples.shape, torch.Size([5, 3, 4, 2]))
```
```
  1) pytorch.botorch.test.models.test_gp_regression.TestSingleTaskGP: ftest_w_none
    1) _LinAlgError: torch.linalg.eigh: (Batch element 0): The algorithm failed to converge because the input matrix is ill-conditioned or has too many repeated eigenvalues (error code: 7).
      File "pytorch/botorch/test/models/test_gp_regression.py", line 390, in ftest_w_none
        samples = posterior.rsample(sample_shape=torch.Size([5]))
      File "botorch/posteriors/gpytorch.py", line 100, in rsample
        samples = self.mvn.rsample(
      File "gpytorch/distributions/multitask_multivariate_normal.py", line 239, in rsample
        samples = super().rsample(sample_shape=sample_shape, base_samples=base_samples)
      File "gpytorch/distributions/multivariate_normal.py", line 181, in rsample
        res = covar.zero_mean_mvn_samples(num_samples) + self.loc.unsqueeze(0)
      File "gpytorch/lazy/lazy_tensor.py", line 2134, in zero_mean_mvn_samples
        covar_root = self.root_decomposition().root
      File "gpytorch/utils/memoize.py", line 59, in g
        return _add_to_cache(self, cache_name, method(self, *args, **kwargs), *args, kwargs_pkl=kwargs_pkl)
      File "gpytorch/lazy/lazy_tensor.py", line 1765, in root_decomposition
        evals, evecs = self.symeig(eigenvectors=True)
      File "gpytorch/utils/memoize.py", line 59, in g
        return _add_to_cache(self, cache_name, method(self, *args, **kwargs), *args, kwargs_pkl=kwargs_pkl)
      File "gpytorch/lazy/lazy_tensor.py", line 1982, in symeig
        return self._symeig(eigenvectors=eigenvectors)
      File "gpytorch/lazy/lazy_tensor.py", line 2314, in _symeig
        evals, evecs = torch.linalg.eigh(self.evaluate().to(dtype=settings._linalg_dtype_symeig.value()))
```

`None` has many more issues than that, so getting that to work would pretty much require rewriting `SingleTaskGP.__init__` (none of the validation, dimension setting or anything that depends on these will work). 

> SingleTaskGP also seems to work well when instantiated with train_X=None, train_Y=None.

@dynobo how did you get this to work?

**Reply by Balandat** [reply]
I mean "posterior" is a funny concept to begin with if there isn't any data observed. The gpytorch model should really only be called in prior mode then.

**Reply by saitcakmak** [reply]
> I mean "posterior" is a funny concept to begin with if there isn't any data observed. The gpytorch model should really only be called in prior mode then.

Agreed. I am just not seeing the motivation for supporting `None` inputs in that case. If the model will not work like any other `SingleTaskGP` and will throw errors here and there, what purpose does this serve? You couldn't start iterations with no data and build up from there since it will error out.
If the goal was to just get the prior, we could initiate it with some dummy data and get the same result (or directly evaluate the `covar_module` without needing a model).

**Reply by Balandat** [reply]
Yeah agreed, we shouldn't go out of our way to support this. But since the model is a gpytorch exact gp under the hood you could just `__call__` it and it would work for that. So if it's easy to do that without much trouble we could consider supporting that.

**Reply by ptighin** [reply]
Let me try to explain why I believe that initializing `SingleTaskGP` with empty data can be super useful. Imagine that we have an informative prior with rich structure (e.g., a non-stationary kernel), which `SingleTaskGP` supports as far as I understand. We would like the acquisition function of Bayesian optimization to tell us which should be the very first point to sample. In other words, the acquisition function should be able to predict on a model with no training data.

You're of course right that one could, in principle, hack around this and write an own wrapper around `SingleTaskGP` to accommodate such a possibility. Supporting this natively, however, may make people's life easier and enable lots of useful use cases.

**Reply by Balandat** [reply]
Yeah that makes sense. I believe that from the gpytorch side this shouldn't really be an issue. But I think we make the assumption that these are not `None` in a lot of places, and so changing that would require some careful audit to see what we need to unbreak (e.g. we grab a lot of info from the training inputs (dimensions, batch shapes etc) to populate some of our default model components in the constructor, and then I am sure we're trying to use some of that information in our `posterior` call as well). So supporting this first-class will require a nontrivial amount of work.

**Reply by ptighin** [reply]
@Balandat I understand your concern. What if we make it work by initialising `SingleTaskGP` with empty tensors, e.g., `train_X = torch.empty(*batch_size, 0, num_features)` and `train_Y = torch.empty(*batch_size, 0, num_outputs)`. Here, the constructor would be able to initialize data-related quantities like batch size etc. If I understand correctly, the only change would need to happen with the normalizers, `input_transform` and `outcome_transform`, which shouldn't be too problematic I think. They could, for instance, be moved inside `SingleTaskGP.condition_on_observations()` and be initialized whenever the first training data comes along. Would something like this be feasible? I would be happy to open a PR on this in case you support the idea.

**Reply by saitcakmak** [reply]
@ptighin See #1229 for a starting point. The changes I made there makes it so that we can initialize `SingleTaskGP` with empty inputs. However, you'll notice that the sampling part of the unit test errors out, which is due to the `posterior.mvn.covariance_matrix` being NaN. See if you can fix that without a major surgery.

Note that I am testing the multi-output `SingleTaskGP` there. This is intentional since BoTorch models treat multiple outputs different than `ExactGP` (the outcome dimension is -1, which requires a bunch of reshaping of tensors). Whatever fix we implement, it should be rather minimal and should not require duplicating a bunch of code across the codebase (to support this with other models etc).

**Reply by ptighin** [reply]
> @ptighin See #1229 for a starting point. The changes I made there makes it so that we can initialize `SingleTaskGP` with empty inputs. However, you'll notice that the sampling part of the unit test errors out, which is due to the `posterior.mvn.covariance_matrix` being NaN. See if you can fix that without a major surgery.
> 
> Note that I am testing the multi-output `SingleTaskGP` there. This is intentional since BoTorch models treat multiple outputs different than `ExactGP` (the outcome dimension is -1, which requires a bunch of reshaping of tensors). Whatever fix we implement, it should be rather minimal and should not require duplicating a bunch of code across the codebase (to support this with other models etc).

Thanks a lot for following up on this. I think that the fix is pretty easy: if empty data is passed, no transforms should be applied to the data since it makes no sense to normalise an empty vector. See the fix at https://github.com/saitcakmak/botorch/commit/dcee8bb432848d9ef4b4c1d7c982000eaf3afac5. What do you think? Can we do a PR with such a solution?

**Reply by ptighin** [reply]
The other option is to forbid the passing of input and output transforms to a model with empty train data. This might give more meaningful feedback to the users that standardisation of an empty array makes little sense.

The third option is to adapt the input and output transforms such that they can handle this corner case under the hood. For instance, `Standardize()` would take a "trivial" state with mean zero and variance one. This option, however, requires a little more surgery I think, since all normalisers would need to be adjusted to this corner case.

What do you think, @saitcakmak?

**Reply by ptighin** [reply]
@saitcakmak I managed to mess up the fork to the above link and your tests disappeared. Please see this commit, https://github.com/saitcakmak/botorch/commit/310246d99f84de0f1375ab59437fbc5b446154f2, for the correct version. Note that I also expanded the tests for empty inputs a little.

**Reply by saitcakmak** [reply]
> if empty data is passed, no transforms should be applied to the data since it makes no sense to normalise an empty vector.

This makes sense on a conceptual level. One concern with this is that we apply transforms to both train and test data. In many cases, we also use the train data to set the parameters of the transforms (this is pretty much why the tests were failing). Suppose you use a prior that is meant to work with standardized data (botorch defaults are designed for standardized data). If you use a transform and pass no training data, your transforms are not properly initialized, and we get around this by not applying them (at least in the proposed design). If we then evaluate some test inputs on this model, these will be in untransformed space. When we use the same prior with non-empty train data, then the test data will be on standardized space. So, we end up with a bit of inconsistency there. To me this just says if you want to use empty inputs with your model, you shouldn't ever use transforms with it - which agrees with your suggestion.

> See the fix at https://github.com/saitcakmak/botorch/commit/dcee8bb432848d9ef4b4c1d7c982000eaf3afac5. What do you think?

I updated #1229 with your changes, and also added the model training step to the unit test, which fails. It makes sense that you should not be able to train a model without any training data. We could probably add a check somewhere in `fit_gpytorch_model` that skips training if there's no data.

A main concern about that design is overwriting the transforms to `None` without warning the user (and the need to do this in each model). Silently overwriting the user inputs is generally a bad practice and we should avoid that. I'd rather let it error out while trying to transform things.

> The third option is to adapt the input and output transforms such that they can handle this corner case under the hood. 

This would require lots of changes in transforms, so ideally we wouldn't do that.

All that said, one cleanish solution I see is just making the change in `transform_inputs` to skip over empty inputs and adding something in `fit_gpytorch_model` to skip training if there are no inputs. This would be rather minimal and would support using empty inputs as long as you don't use any transforms (you can always manually transform the data in your code). Would that work for you? I can look into what needs to be done in `fit_gpytorch_model` to make this work.

**Reply by saitcakmak** [reply]
Actually, you don't even need the change in `transform_inputs` if the `input_transform is None`. So, we would only need to skip model fitting

**Reply by saitcakmak** [reply]
Updated #1229 with the proposed change to `fit_gpytorch_model`.
